plutus-core/plutus-core/src/PlutusCore/Builtin/Result.hs

```diff
@@ -181,6 +188,10 @@ throwNotAConstant :: MonadError BuiltinError m => m void
 throwNotAConstant = throwing _StructuralUnliftingError "Not a constant"
 {-# INLINE throwNotAConstant #-}
```

Comment:
Does inlining things like this make any difference to general performance?  Presumably this is something that's only going to happen once, so is optimising it helpful?  I may well be missing something important though.

Reply:
> Does inlining things like this make any difference to general performance?

Yes, it can make it worse :upside_down_face: 

I'm doing it, because

1. I read GHC Core for builtins a lot and I want everything in there look as bare-bones as possible, otherwise I start thinking "should this be optimized?" and get distracted. Without the pragma the function would likely still be inlined, but if it wasn't, we'd get a constraint dictionary in the middle of builtins and that's one thing that I don't want to see in there
2. it's much less likely that somebody forgets an `INLINE` pragma in an important place if we simply have them everywhere as that is an easier rule to follow than have `INLINE` here and `NOINLINE` here
3. if inlining this definition hurts performance, then it's a bug elsewhere where we don't have an appropriate `oneShot` call or something and that bug then can be triggered by other code, so exposing it's here with simple error throwing is kinda a feature

I'll turn this into a Note, thanks for the question.

Reply:
Turned out to be quite a rabbit hole. I've added this Note:

```
{- Note [INLINE and OPAQUE on error-related definitions]
We mark error-related definitions such as prisms like '_StructuralUnliftingError' and regular
functions like 'throwNotAConstant' with @INLINE@, because this produces significantly less cluttered
GHC Core. Not doing so results in 20+% larger Core for builtins.

However in a few specific cases we use @OPAQUE@ instead to get tighter Core. @OPAQUE@ is the same as
@NOINLINE@ except the former _actually_ prevents GHC from inlining the definition unlike the latter.
See this for details: https://github.com/ghc-proposals/ghc-proposals/blob/5577fd008924de8d89cfa9855fa454512e7dcc75/proposals/0415-opaque-pragma.rst

It's hard to predict where @OPAQUE@ instead of @INLINE@ will help to make GHC Core tidier, so it's
mostly just looking into the Core and seeing where there's obvious duplication that can be removed.
Such cases tend to be functions returning a value of a concrete error type (as opposed to a type
variable).
-}
```

If we use `NOINLINE` instead of `INLINE` in those places, then we'll get ~22k lines of Core instead of ~18k. But using `INLINE` and just a couple of well-placed `OPAQUE` calls we get the number of lines down to less than 15k.

I'll also bring that `OPAQUE` thing to the team w.r.t. Plutus Tx, since it's way more reliable than `NOINLINE` that we user currently.